typescript/testsuite/network_impl/test_network.ts

```diff
@@ -0,0 +1,350 @@
+
+        let personRetrievingApiClientResult: Result<APIClient, Error>;
+        try {
+            personRetrievingApiClientResult = await this.addApiService();
+        } catch(exception: any) {
+            // Sadly, we have to do this because there's no great way to enforce the caught thing being an error
+            // See: https://stackoverflow.com/questions/30469261/checking-for-typeof-error-in-js
+            if (exception && exception.stack && exception.message) {
+                return err(exception as Error);
+            }
+            return err(new Error("Calling addApiService method on our TestNetwork class threw an exception, but " +
+                "it's not an Error so we can't report any more information than this"));
+        }
+        if (!personRetrievingApiClientResult.isOk()) {
+            return err(personRetrievingApiClientResult.error);
+        }
+        this.personRetrievingApiClient = personRetrievingApiClientResult.value;
+
+        return ok(null);
+    }
+
+    //  Custom network implementations will also usually have getters, to retrieve information about the
+    //  services created during setup
+    public getPersonModifyingApiClient(): Result<APIClient, Error> {
+        if (this.personModifyingApiClient === null) {
+            return err(new Error("No person-modifying API client exists"));
+        }
+        return ok(this.personModifyingApiClient);
+    }
+
+    public getPersonRetrievingApiClient(): Result<APIClient, Error> {
+        if (this.personRetrievingApiClient === null) {
+            return err(new Error("No person-retrieving API client exists"));
+        }
+        return ok(this.personRetrievingApiClient)
+    }
+
+    public getDatastoreClient(): Result<DatastoreClient, Error>{
+        if (this.datastoreClient === null) {
+            return err(new Error("No datastore client exists"));
+        }
+        return ok(this.datastoreClient);
+    }
+
+    // ====================================================================================================
+    //                                       Private helper functions
+    // ====================================================================================================
+    
+    private async addApiService(): Promise<Result<APIClient, Error>> {
+
+        if (this.datastoreClient === null) {
+            return err(new Error("Cannot add API service to network; no datastore client exists"));
+        }
+    
+        const serviceIdStr: string = API_SERVICE_ID_PREFIX + this.nextApiServiceId.toString();
+        this.nextApiServiceId = this.nextApiServiceId + 1;
+        const serviceId: ServiceID = <ServiceID>(serviceIdStr);
+    
+        let configInitializingFunc: (fp: number) => Promise<Result<null, Error>>;
+        try {
+            configInitializingFunc = await TestNetwork.getApiServiceConfigInitializingFunc(this.getDatastoreClient());
```

Comment:
You don't need to:
1) use `this.getDatastoreClient()` - you can just use `this.datastoreClient` directly, because you already check that it's non-null as the first line of this function
2) we shouldn't ever pass a `Result<>` as part of function args anyways - the `Result` error checking should be immediately after getting it always, which prevents us from forgetting it